src/main/kotlin/uk/gov/dluhc/printapi/service/CertificateBatchingService.kt

```diff
@@ -42,10 +42,7 @@ class CertificateBatchingService(
 
         return limitCertificates(certificatesPendingAssignment).chunked(batchSize).associate { batchOfCertificates ->
             val batchId = idFactory.batchId()
-            batchId to batchOfCertificates.onEach {
-                it.getCurrentPrintRequest().batchId = batchId
-                it.addStatus(ASSIGNED_TO_BATCH)
-            }
+            batchId to batchOfCertificates.onEach { it.addPrintRequestToBatch(batchId) }
```

Comment:
In the above code when querying certificates by status, we might miss certificates which had a new print request with `PENDING_ASSIGNMENT_TO_BATCH`, but then received a response for a prior print request, assume `VALIDATED_BY_PRINT_PROVIDER`, which will update the status of certificate to `VALIDATED_BY_PRINT_PROVIDER`, hence not being returned by the `certificateRepository.findByStatusOrderByApplicationReceivedDateTimeAsc` method.

Reply:
There is also one more issue here. Assume somehow we have one `printRequest` with status `PENDING_ASSIGNMENT_TO_BATCH` and before it being assigned to a batch we receive a new `printRequest` and its status also becomes `PENDING_ASSIGNMENT_TO_BATCH`.
With the `it.addPrintRequestToBatch(batchId)`, we will be assigning both of them to the same batch, but it will count as one in our limiting logic, since we iterate over certificates and assume we assign one request per certificate to the batch (which was true for the previous implementation), but that might no longer be true.

Reply:
@Neil, We might need to change the logic here as well in the `ProcessPrintBatchService.processBatch`.

<pre>
 @Transactional
    fun processBatch(batchId: String) {
        val certificates = certificateRepository.findByStatusAndPrintRequestsBatchId(ASSIGNED_TO_BATCH, batchId)
        if (certificates.isEmpty()) {
            throw EmptyBatchException(batchId, ASSIGNED_TO_BATCH)
        }
        val fileContents = printFileDetailsFactory.createFileDetailsFromCertificates(batchId, certificates)
        val sftpInputStream = sftpZipInputStreamProvider.createSftpInputStream(fileContents)
        val sftpFilename = filenameFactory.createZipFilename(batchId, certificates.size)
        sftpService.sendFile(sftpInputStream, sftpFilename)
        updateCertificates(batchId, certificates)
    }

</pre>

We are querying by the status in the Certificate, which might have a `ASSIGNED_TO_BATCH` print request, but that might not be status of the Certificate if there was a newer print request added to that certificate, which is not assigned to a batch yet. 

I think we need to query print requests directly from the printRequestRepository based on their status.

Also when generating file details with the  `printFileDetailsFactory.createFileDetailsFromCertificates` we need to take into account that there might be more than one print request for a certificate.

<pre>
  fun createFileDetailsFromCertificates(batchId: String, certificates: List<Certificate>): FileDetails {
        val fileContents = createFromCertificates(certificates)
        return FileDetails(
      </b>      printRequestsFilename = filenameFactory.createPrintRequestsFilename(batchId, certificates.size),</b>
            printRequests = fileContents.printRequests,
            photoLocations = fileContents.photoLocations
        )
    }

</pre>

Reply:
`certificateRepository.findByStatusOrderByApplicationReceivedDateTimeAsc` probably should be refactored as it only handles getting `PENDING_ASSIGNMENT_TO_BATCH` and that's the first status - the code I think is fine and the situation you describe I don't think is possible.

Reply:
```
With the it.addPrintRequestToBatch(batchId) code is executed, we will be assigning both of them to the same batch, but it will count as one in our limiting logic, since we iterate over certificates and assume we assign one request per certificate to the batch (which was true for the previous implementation), but that might no longer be true.
```
Yes, I had considered this as an extreme edge case as when discussed with Nathan there was no expectation that the Admin will be making multiple re-print requests within a batching window.  (In effect we expect reprints to be occasional and not fired at the system at a high rate). If this does turn out to be a real issue the simplest solution is to reduce the threshold until the limiting code is fixed.

cc  @NathanRussellValtech - in case you think this needs to be included as part of this ticket

Reply:
Thanks, @Neil-Massey , for the huddle we had. 

✅ For the first issue I mentioned, I agree that this won't actually happen and I raised it because I misinterpreted the status calculation on the certificate. So it is all good.

✅ For the second one about having 2 requests assigned to a batch, I agree that exceeding the limit because of this is a very extreme case and since the implementation to prevent it is also complicated, I am fine with the current way we do it.

✅ For the third one about `ProcessPrintBatchService.processBatch`, you addressed it as well. Thanks.

Lastly, about the `printRequestsFilename = filenameFactory.createPrintRequestsFilename(batchId, certificates.size)`. Seems like it will be an easy fix to calculate the request size for sending to print by going through the print requests in the certificates with the correct status.

<pre>
    fun createFileDetailsFromCertificates(batchId: String, certificates: List<Certificate>): FileDetails {
        val fileContents = createFromCertificates(certificates)
        val sizeOfAssignedToBatchPrintRequests = certificates.flatMap { it.printRequests }
            .filter { it.getCurrentStatus().status == Status.ASSIGNED_TO_BATCH }.size
        return FileDetails(

            printRequestsFilename = filenameFactory.createPrintRequestsFilename(batchId, sizeOfAssignedToBatchPrintRequests),
            printRequests = fileContents.printRequests,
            photoLocations = fileContents.photoLocations
        )
    }
</pre>

Thanks, again.

Reply:
Refactored to filename factory 👍